src/mpid/ch4/shm/ipc/gpu/gpu_post.h

```diff
@@ -57,15 +62,20 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_GPU_get_mem_attr(const void *vaddr, MPIDI_IPC
 MPL_STATIC_INLINE_PREFIX int MPIDI_GPU_attach_mem(MPIDI_GPU_mem_handle_t mem_handle,
                                                   MPL_gpu_device_handle_t dev_handle, void **vaddr)
 {
+    int remote_dev_id;
+    MPL_gpu_device_handle_t remote_dev_handle;
     int mpi_errno = MPI_SUCCESS, mpl_err = MPL_SUCCESS;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_GPU_ATTACH_MEM);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_GPU_ATTACH_MEM);
 
-    if (mem_handle.dev_id == -1)
+    if (mem_handle.global_dev_id == -1)
         mpl_err = MPL_gpu_ipc_handle_map(mem_handle.ipc_handle, dev_handle, vaddr);
     else {
         MPL_gpu_device_handle_t remote_dev_handle;
-        MPL_gpu_get_dev_handle(mem_handle.dev_id, &remote_dev_handle);
+        MPIDI_gpu_dev_id_t *tmp;
+        HASH_FIND_INT(MPIDI_gpu_global.global_to_local_map, &mem_handle.global_dev_id, tmp);
+
+        MPL_gpu_get_dev_handle(tmp->local_dev_id, &remote_dev_handle);
```

Comment:
Except for `MPIDI_GPU_get_mem_attr`, the remaining functions in this file are not performance critical.  So they don't need to be implemented in a header file.  It's better to move them to a C file, so the internal data structures it uses are not exposed to the IPC layer.